### PR TITLE
Add manual Snap update feature from Snap View component

### LINKED ISSUE #21563: Update from settings
Enables users to update to the latest version of their snap from within the MetaMask client.

Design: https://www.figma.com/file/3Z2k9kDpHxoyjwtQo1DAam/Home?type=design&node-id=1%3A35224&mode=design&t=h2KhzVnz1zW1nXRo-1

### PR DESCRIPTION
**Pull request by david0xd**
## **Description**
This PR adds new feature for updating Snap from Snap View component (inside extension). Update can be triggered manually, when available update banner alert is displayed.

## **Related issues**
Fixes: https://github.com/MetaMask/metamask-extension/issues/21563

## **Manual testing steps**
1. Install a test Snap from older versions (e.g. https://metamask.github.io/snaps/test-snaps/0.38.0-flask.1/)
3. Go to Snaps List page and check if snap has a dot that indicates the update is available
4. Select the Snap that has an update indicator
5. Check if Snap Alert Banner is displayed
6. Click on `Update` button-link inside the banner
7. Proceed with the Snap update following the update popups
8. After Snap has been successfully updated, dot indicator or update alert banner show not be present for a given snap.

### **Before**
Before changes in this PR update flow didn't exist. So, nothing special to screenshot.

### **After**
Screenshot 2023-11-14 at 00 06 28
Screenshot 2023-11-14 at 00 07 50
Screenshot 2023-11-14 at 00 06 53
Screenshot 2023-11-14 at 00 07 27
Screenshot 2023-11-16 at 15 27 23
Screenshot 2023-11-16 at 15 44 20
Screenshot 2023-11-16 at 15 44 55

## **Pre-merge author checklist**
- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**
- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/store/actions.ts [hunk 1]
```diff
@@ -1136,6 +1137,22 @@ export function enableSnap(
   };
 }
 
+export function updateSnap(
+  origin: string,
+  snap: { [snapId: string]: { version: string } },
+): ThunkAction<void, MetaMaskReduxState, unknown, AnyAction> {
+  return async (dispatch: MetaMaskReduxDispatch, getState) => {
+    submitRequestToBackground('installSnap', [origin, snap]);
+    await forceUpdateMetamaskState(dispatch);
```
Comment (FrederikBolding) [inline ui/store/actions.ts:1146]: `forceUpdateMetamaskState` returns the new state, so I think we can simplify this

File: ui/selectors/selectors.js [hunk 2]
```diff
@@ -718,6 +717,60 @@ export function getSnapRegistryData(state, snapId) {
...
 
+/**
+ * Find and return Snap's latest version available in registry.
+ *
+ * @param state - Redux state object.
+ * @param snapId - ID of a Snap.
+ * @returns String SemVer version.
+ */
+export function getSnapLatestVersion(state, snapId) {
+  const snapRegistryData = getSnapRegistryData(state, snapId);
```
Comment (FrederikBolding) [inline ui/selectors/selectors.js:728]: I wonder if we shouldn't be creating reselect selectors here instead?

File: ui/pages/permissions-connect/snaps/snap-update/snap-update.js [hunk 3]
```diff
@@ -148,14 +146,6 @@ export default function SnapUpdate({
               textAlign={TEXT_ALIGN.CENTER}
             >
               {t('snapUpdateRequest', [
-                <Text
-                  as={ValidTag.Span}
-                  key="1"
-                  variant={TextVariant.bodyMd}
-                  fontWeight={FontWeight.Medium}
-                >
-                  {originMetadata?.hostname}
```
Comment (FrederikBolding) [inline ui/pages/permissions-connect/snaps/snap-update/snap-update.js:157]: Are we sure we want to strip this? I would assume we only strip the origin when the origin is MM?

cc @eriknson

File: ui/components/app/snaps/snap-update-alert/snap-update-alert.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+      title="Update available"
+      actionButtonLabel="Update"
+      actionButtonOnClick={onUpdateClick}
+      actionButtonProps={{
+        endIconName: IconName.Download,
+        color: TextColor.primaryDefault,
+      }}
+      {...bannerAlertProps}
+    >
+      <Text>Get the latest version of {snapName}</Text>
```
Comment (FrederikBolding) [inline ui/components/app/snaps/snap-update-alert/snap-update-alert.js:18]: All of the text in this file should be translatable

File: ui/selectors/selectors.js [hunk 5]
```diff
@@ -718,6 +717,60 @@ export function getSnapRegistryData(state, snapId) {
...
+
+/**
+ * Return status of Snaps update availability for any installed Snap.
+ *
+ * @param state - Redux state object.
+ * @returns boolean true if update is available, false otherwise.
+ */
+export function getSnapsUpdateAvailableStatus(state) {
+  const snapMap = getAllSnapsAvailableUpdateStatus(state);
+  return [...snapMap.values()].some((value) => value === true);
```
Comment (FrederikBolding) [inline ui/selectors/selectors.js:771]: Do we need to spread here?

**Review by FrederikBolding — CHANGES_REQUESTED** [review]

File: ui/selectors/selectors.js (hunk 5 above)
Comment (david0xd) [reply ui/selectors/selectors.js:771]: I think yes if I want to use `.some`, I need a regular array returned with the values from map 🤔

File: ui/components/app/snaps/snap-update-alert/snap-update-alert.js (hunk 4 above)
Comment (david0xd) [reply ui/components/app/snaps/snap-update-alert/snap-update-alert.js]: Thanks for noticing! I forgot about these. Fixed now.

File: ui/store/actions.ts (hunk 1 above)
Comment (david0xd) [reply ui/store/actions.ts:1146]: If I do so, the UX becomes broken again. So, the popup with the update screen will remain in the background. I think that @GuillaumeRx have strong reasons why this is all implemented like that.

Comment (FrederikBolding) [reply ui/store/actions.ts:1146]: Yeah, you are right. I think this has a race condition. We may want to find a way to wait for the approval to trigger. Currently I think we could be unlucky at get stuck.

File: ui/selectors/selectors.js (hunk 2 above)
Comment (david0xd) [reply ui/selectors/selectors.js]: Refactored everything to use memoized selectors with `createSelector`. I'm just not sure if using `useSelector` to pass the state to it is the best approach, but I do not see other way for passing the state in some established way from these components (let me know if you think there is something better).

File: app/scripts/metamask-controller.js [hunk 6]
```diff
@@ -2944,6 +2944,20 @@ export default class MetamaskController extends EventEmitter {
         this.controllerMessenger,
         'SnapController:enable',
       ),
+      installSnap: this.controllerMessenger.call.bind(
```
Comment (FrederikBolding) [inline app/scripts/metamask-controller.js:2947]: I don't think this is used anymore

File: ui/selectors/selectors.js [hunk 7]
```diff
@@ -706,17 +705,117 @@ export function getTargetSubjectMetadata(state, origin) {
...
+  },
+);
+
+/**
+ * Return status of Snaps update availability for any installed Snap.
+ *
+ * @param state - Redux state object.
+ * @returns boolean true if update is available, false otherwise.
+ */
+export const getSnapsUpdateAvailableStatus = createSelector(
```
Comment (FrederikBolding) [inline ui/selectors/selectors.js:812]:
```suggestion
export const getAnySnapUpdateAvailable = createSelector(
```
I think this is slightly more clear?

File: ui/selectors/selectors.js [hunk 8]
```diff
@@ -706,17 +705,117 @@ export function getTargetSubjectMetadata(state, origin) {
...
+  },
+);
+
+/**
+ * Return a Map of all installed Snaps with available update status.
+ *
+ * @param state - Redux state object.
+ * @returns Map Snap IDs mapped to a boolean value (true if update is available, false otherwise).
+ */
+export const getAllSnapsAvailableUpdateStatus = createSelector(
```
Comment (FrederikBolding) [inline ui/selectors/selectors.js:786]:
```suggestion
export const getAllSnapAvailableUpdates = createSelector(
```
I think this is slightly more clear?

File: ui/pages/snaps/snap-view/snap-settings.js [hunk 9]
```diff
@@ -146,8 +152,37 @@ function SnapSettings({ snapId }) {
     setIsDescriptionOpen(true);
   };
 
+  const snapLatestVersion = useSelector((state) =>
```
Comment (FrederikBolding) [inline ui/pages/snaps/snap-view/snap-settings.js:155]:
```suggestion
  const latestRegistryVersion = useSelector((state) =>
```
Maybe?

File: ui/pages/snaps/snap-view/snap-settings.js [hunk 10]
```diff
@@ -146,8 +152,37 @@ function SnapSettings({ snapId }) {
...
+    snap ? getSnapLatestVersion(state, snap?.id) : null,
+  );
+  const checkAvailableUpdate = () => {
```
Comment (FrederikBolding) [inline ui/pages/snaps/snap-view/snap-settings.js:158]: Do we need a function here? Could we not just do:
```
const isUpdateAvailable = snapLatestVersion
      ? semver.gt(snapLatestVersion, snap.version)
      : false
```

**Review by FrederikBolding — CHANGES_REQUESTED** [review]

File: ui/pages/snaps/snap-view/snap-settings.js (hunk 10 above)
Comment (david0xd) [reply ui/pages/snaps/snap-view/snap-settings.js]: Updated / refactored.

File: ui/selectors/selectors.js (hunk 8 above)
Comment (david0xd) [reply ui/selectors/selectors.js]: Updated / refactored.

File: ui/selectors/selectors.js (hunk 7 above)
Comment (david0xd) [reply ui/selectors/selectors.js]: Updated / refactored.

File: app/scripts/metamask-controller.js (hunk 6 above)
Comment (david0xd) [reply app/scripts/metamask-controller.js]: Removed.

File: ui/selectors/selectors.js [hunk 11]
```diff
@@ -706,17 +705,117 @@ export function getTargetSubjectMetadata(state, origin) {
...
+export const rawStateSelector = (state) => state;
+
+/**
+ * Input selector used to retrieve Snaps that are added to Snaps Directory.
+ *
+ * @param state - Redux state object.
+ * @returns Object - Containing verified Snaps from the Directory.
+ */
+const selectVerifiedSnapsRegistry = (state) =>
+  state.metamask.database.verifiedSnaps;
```
Comment (FrederikBolding) [inline ui/selectors/selectors.js:727]:
```suggestion
  state.metamask.database?.verifiedSnaps;
```
I don't think this is safe, `state.metamask.database` is `null` by default.

File: ui/pages/snaps/snap-view/snap-settings.js [hunk 12]
```diff
@@ -146,8 +152,34 @@ function SnapSettings({ snapId }) {
     setIsDescriptionOpen(true);
   };
 
+  const latestRegistryVersion = useSelector((state) =>
+    snap ? getSnapLatestVersion(state, snap?.id) : null,
+  );
+  const isUpdateAvailable = latestRegistryVersion
```
Comment (FrederikBolding) [inline ui/pages/snaps/snap-view/snap-settings.js:158]:
```suggestion

  const isUpdateAvailable = latestRegistryVersion
```

File: ui/pages/snaps/snap-view/snap-settings.js [hunk 13]
```diff
@@ -146,8 +152,34 @@ function SnapSettings({ snapId }) {
...
+    ? semver.gt(latestRegistryVersion, snap.version)
+    : false;
+  const handleUpdate = async () => {
```
[the remark this answers is not in the document]
Comment (david0xd) [reply ui/pages/snaps/snap-view/snap-settings.js:163]: Updated.

File: ui/selectors/selectors.js (hunk 11 above)
Comment (david0xd) [reply ui/selectors/selectors.js]: Updated.

File: ui/pages/snaps/snap-view/snap-settings.js (hunk 12 above)
Comment (david0xd) [reply ui/pages/snaps/snap-view/snap-settings.js:159]: Updated.

**Review by FrederikBolding — DISMISSED** [review]

**Review by FrederikBolding — DISMISSED** [review]

File: ui/components/multichain/global-menu/global-menu.js [hunk 14]
```diff
@@ -92,6 +93,9 @@ export const GlobalMenu = ({ closeMenu, anchorElement }) => {
 
   ///: BEGIN:ONLY_INCLUDE_IN(snaps)
   const unreadNotificationsCount = useSelector(getUnreadNotificationsCount);
+  const snapsUpdatesAvailable = useSelector((state) =>
+    getAnySnapUpdateAvailable(state),
+  );
```
Comment (GuillaumeRx) [inline ui/components/multichain/global-menu/global-menu.js:98]:
```suggestion
  const snapsUpdatesAvailable = useSelector(getAnySnapUpdateAvailable);
```

File: ui/components/ui/menu/menu-item.js [hunk 15]
```diff
@@ -25,7 +38,25 @@ const MenuItem = React.forwardRef(
...
+              name={IconName.FullCircle}
+              size={IconSize.Xs}
+              color={IconColor.primaryDefault}
+              style={{ '--size': '10px' }}
+            />
+          }
+        >
+          <Icon name={iconName} size={IconSize.Sm} marginRight={2} />
+        </BadgeWrapper>
+      ) : null}
```
Comment (GuillaumeRx) [inline ui/components/ui/menu/menu-item.js:58]:
```suggestion
      {iconName && showInfoDot && (
        <BadgeWrapper
          anchorElementShape={BadgeWrapperAnchorElementShape.circular}
          display={Display.Block}
          position={BadgeWrapperPosition.topRight}
          positionObj={{ top: -6, right: 4 }}
          badge={
            <Icon
              name={IconName.FullCircle}
              size={IconSize.Xs}
              color={IconColor.primaryDefault}
              style={{ '--size': '10px' }}
            />
          }
        >
          <Icon name={iconName} size={IconSize.Sm} marginRight={2} />
        </BadgeWrapper>
      )}
```

File: ui/pages/snaps/snaps-list/snap-list.js [hunk 16]
```diff
@@ -51,6 +51,9 @@ const SnapList = () => {
   }, [settingsRef, t]);
 
   const snapsList = useSelector((state) => getSnapsList(state));
+  const snapUpdateMap = useSelector((state) =>
+    getAllSnapAvailableUpdates(state),
+  );
```
Comment (GuillaumeRx) [inline ui/pages/snaps/snaps-list/snap-list.js:56]:
```suggestion
  const snapsList = useSelector(getSnapsList);
  const snapUpdateMap = useSelector(getAllSnapAvailableUpdates);
```

**Review by GuillaumeRx — CHANGES_REQUESTED** [review]

File: ui/components/multichain/global-menu/global-menu.js (hunk 14 above)
Comment (david0xd) [reply ui/components/multichain/global-menu/global-menu.js]: Updated.

File: ui/components/ui/menu/menu-item.js (hunk 15 above)
Comment (david0xd) [reply ui/components/ui/menu/menu-item.js]: Updated.

File: ui/pages/snaps/snaps-list/snap-list.js (hunk 16 above)
Comment (david0xd) [reply ui/pages/snaps/snaps-list/snap-list.js]: Updated.

**Review by GuillaumeRx — APPROVED** [review]

**Review by FrederikBolding — APPROVED** [review]

**Pull request merged**